File: web/src/app/recipe/schema.ts
export const Units = ["cup", "tsp", "Tbsp", "lbs", "kg", "g", "mL", "L", "pt", "pt", "gal", ""]
export const ModifierDividerRegex = /[,;]\s*/

export type RecipeModel = {
  uid: string,
  created_at: string,
  updated_at: string,
  is_trashed: boolean,
  user_uid: string,
  name: string,
  description: string,
  meal: string,
  instructions: string,
  ingredients: IngredientModel[],
}

export type IngredientModel = {
  uid: string,
  user_uid: string,
  recipe_uid: string,
  created_at: string,
  updated_at: string,
  meal: string,
  order: number,
  unit: string | null,
  quantity: string | number,
  item: string,
  optional: boolean,
  modifier: string | null,
}

export const MealTypes = new Map<string, string>([
  ["breakfast", "Breakfast"],
  ["lunch", "Lunch"],
  ["dinner", "Dinner"],
  ["dessert", "Dessert"],
  ["none", "None"],
])

export interface UpdateRecipe {
  name?: string
  meal?: string
  description?: string
  instructions?: string
}
export interface UpdateIngredient {
  quantity?: string,
  unit?: string,
  item?: string,
  optional?: boolean
  modifier?: string | null
}

export function CreateRecipeRequest(model: RecipeModel): any {
  return {
    name: model.name,
    description: model.description,
    meal: model.meal,
    instructions: model.instructions,
    ingredients: model.ingredients.map(i => {
      return {
        quantity: i.quantity, unit: i.unit, item: i.item, modifier: i.modifier, optional: i.optional
      }
    }),
  }
}

export function UpdateRecipeRequest(model: RecipeModel): any {
  return {
    uid: model.uid,
    name: model.name,
    is_trashed: model.is_trashed,
    description: model.description,
    meal: model.meal,
    ingredients: model.ingredients.map(i => {
      return {
        quantity: i.quantity, unit: i.unit, item: i.item, modifier: i.modifier, optional: i.optional
      }
    }),
    instructions: model.instructions,
  }
}

File: core/services/recipe/schema.go
package recipe

import (
	"cardamom/core/ext/log_ext"
	m "cardamom/core/ext/math_ext"
	u "cardamom/core/ext/units"
	"cardamom/core/models"
	"strings"

	"github.com/thoas/go-funk"
)

type IngredientPart struct {
	Quantity m.Rational `json:"quantity"`
	Unit     *u.Unit    `json:"unit"`
	Item     string     `json:"item"`
	Optional bool       `json:"optional"`
	Modifier string     `json:"modifier"`
}

type CreateRecipeRequest struct {
	Name         string           `json:"name"`
	Description  string           `json:"description,omitempty"`
	Meal         models.MealType  `json:"meal"`
	Ingredients  []IngredientPart `json:"ingredients"`
	Instructions string           `json:"instructions"`
}

func (req *CreateRecipeRequest) Validate() (string, error) {

	req.Name = strings.TrimSpace(req.Name)
	if len(req.Name) == 0 {
		return log_ext.ReturnBoth("name must not be empty in request")
	}

	if len(req.Ingredients) == 0 {
		return log_ext.ReturnBoth("ingredients must not be empty in request")
	}

	if len(req.Instructions) == 0 {
		return log_ext.ReturnBoth("instructions must not be empty in request")
	}

	for i, ingre := range req.Ingredients {
		req.Ingredients[i].Item = strings.TrimSpace(strings.ToLower(ingre.Item))
		req.Ingredients[i].Modifier = strings.TrimSpace(strings.ToLower(ingre.Modifier))
	}

	return "", nil
}

type UpdateRecipeRequest struct {
	Uid          string           `json:"uid"`
	IsTrashed    *bool            `json:"is_trashed,omitempty"`
	Name         *string          `json:"name,omitempty"`
	Description  *string          `json:"description,omitempty"`
	Meal         *models.MealType `json:"meal,omitempty"`
	Instructions *string          `json:"instructions,omitempty"`
}

func (req *UpdateRecipeRequest) Validate() (string, error) {

	req.Uid = strings.TrimSpace(req.Uid)
	if len(req.Uid) == 0 {
		return log_ext.ReturnBoth("uid must not be empty in request")
	}

	if req.Name != nil {
		*req.Name = strings.TrimSpace(*req.Name)
		if len(*req.Name) == 0 {
			return log_ext.ReturnBoth("name must not be empty")
		}
	}

	return "", nil
}

type CreateRecipeIngredientRequest struct {
	RecipeUid string     `json:"recipe_uid"`
	Quantity  m.Rational `json:"quantity,omitempty"`
	Unit      *u.Unit    `json:"unit,omitempty"`
	Item      string     `json:"item,omitempty"`
	SortOrder int        `json:"order"`
	Optional  bool       `json:"optional,omitempty"`
	Modifier  string     `json:"modifier,omitempty"`
}

func (req *CreateRecipeIngredientRequest) Validate() (string, error) {

	req.RecipeUid = strings.TrimSpace(req.RecipeUid)
	if len(req.RecipeUid) == 0 {
		return log_ext.ReturnBoth("uid must not be empty in request")
	}
	req.Item = strings.TrimSpace(strings.ToLower(req.Item))
	req.Modifier = strings.TrimSpace(strings.ToLower(req.Modifier))

	return "", nil
}

type UpdateRecipeIngredientRequest struct {
	Uid      string      `json:"uid"`
	Quantity *m.Rational `json:"quantity,omitempty"`
	Unit     *u.Unit     `json:"unit,omitempty"`
	Item     *string     `json:"item,omitempty"`
	Optional *bool       `json:"optional,omitempty"`
	Modifier *string     `json:"modifier,omitempty"`
}

func (req *UpdateRecipeIngredientRequest) Validate() (string, error) {

	req.Uid = strings.TrimSpace(req.Uid)
	if len(req.Uid) == 0 {
		return log_ext.ReturnBoth("uid must not be empty in request")
	}

	if req.Item != nil {
		*req.Item = strings.TrimSpace(strings.ToLower(*req.Item))
	}

	if req.Modifier != nil {
		*req.Modifier = strings.TrimSpace(strings.ToLower(*req.Modifier))
	}

	return "", nil
}

type ReorderRecipeIngredientRequest struct {
	Uid         string   `json:"uid"`
	Ingredients []string `json:"ingredient_uids"`
}

func (req *ReorderRecipeIngredientRequest) Validate() (string, error) {

	req.Uid = strings.TrimSpace(req.Uid)
	if len(req.Uid) == 0 {
		return log_ext.ReturnBoth("uid must not be empty in request")
	}
	return "", nil
}

type SearchRecipeRequest struct {
	Name        *string          `json:"name,omitempty"`
	Description *string          `json:"description,omitempty"`
	Meal        *models.MealType `json:"meal,omitempty"`
	Ingredient  *string          `json:"ingredient,omitempty"`
}

func (req *SearchRecipeRequest) Validate() (string, error) {

	if req.Name == nil && req.Description == nil && req.Meal == nil && req.Ingredient == nil {
		return log_ext.ReturnBoth("must have at least one search criteria")
	}

	return "", nil
}

type GetAvailableRecipeRequest struct {
	Meal *models.MealType `json:"meal,omitempty"`
}

func (req *GetAvailableRecipeRequest) Validate() (string, error) {

	if req.Meal != nil && !funk.Contains(models.ValidMeals, *req.Meal) {
		return log_ext.ReturnBoth("invalid meal type")
	}

	return "", nil
}
